src/main/java/com/minecolonies/entity/ai/basic/AbstractEntityAIBasic.java

```diff
@@ -736,27 +740,104 @@ private void clearActionsDone()
+     * @param i the iteration inside the inventory.
+     * @return false if should be kept.
+     */
+    private boolean shouldKeep(Map<ItemStorage, Integer> keptX, Map<ItemStorage, Integer> toKeep, AbstractBuildingWorker buildingWorker, ItemStack stack, int i)
+    {
+        @Nullable ItemStack returnStack;
+        int amountToKeep = 0;
+        if(keptEnough(keptX, toKeep, stack))
+        {
+            returnStack = InventoryUtils.setStack(buildingWorker.getTileEntity(), stack);
+        }
+        else
+        {
+            final ItemStorage tempStorage = new ItemStorage(stack.getItem(), stack.getItemDamage(), stack.stackSize, false);
+            ItemStack tempStack = handleKeepX(keptX, toKeep, tempStorage);
+            if(tempStack == null || tempStack.stackSize == 0)
+            {
+                return false;
+            }
+            amountToKeep = toKeep.get(tempStorage) - tempStack.stackSize;
+            returnStack = InventoryUtils.setStack(buildingWorker.getTileEntity(), tempStack);
+        }
+        if (returnStack == null)
+        {
+            worker.getInventoryCitizen().decrStackSize(i, stack.stackSize - amountToKeep);
+            return amountToKeep == 0;
+        }
+        worker.getInventoryCitizen().decrStackSize(i, stack.stackSize - returnStack.stackSize - amountToKeep);
+        //Check that we are not inserting into a full inventory.
+        return stack.stackSize != returnStack.stackSize;
+    }
+
+    /**
+     * Handle the cases when X items should be kept.
+     * @param keptX already kept items.
+     * @param toKeep to keep items.
+     * @param tempStorage item to analyze.
+     * @return null if should be kept entirely, else itemStack with amount which should be dumped.
+     */
+    private static ItemStack handleKeepX(Map<ItemStorage, Integer> keptX, Map<ItemStorage, Integer> toKeep, ItemStorage tempStorage)
+    {
+        int amountKept = 0;
+        if(keptX.get(tempStorage) != null)
+        {
+            amountKept = keptX.remove(tempStorage);
+        }
+
+        if (toKeep.get(tempStorage) >= (tempStorage.getAmount() + amountKept))
+        {
+            keptX.put(tempStorage, tempStorage.getAmount() + amountKept);
+            return null;
+        }
+        keptX.put(tempStorage, toKeep.get(tempStorage));
+        int dump = tempStorage.getAmount() + amountKept - toKeep.get(tempStorage);
+
+        //Create tempStack with the amount of items that should be dumped.
+        return new ItemStack(tempStorage.getItem(), dump, tempStorage.getDamageValue());
+    }
+
+    /**
+     * Checks if enough items have been kept already.
+     * @param kept kept items.
+     * @param keep items to keep.
+     * @param stack stack to analyse.
+     * @return true if the the item shouldn't be kept.
+     */
+    private static boolean keptEnough(Map<ItemStorage, Integer> kept, Map<ItemStorage, Integer> keep, ItemStack stack)
+    {
+        for(ItemStorage tempStack: keep.keySet())
```

Comment:
![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png) Iterate over the "entrySet" instead of the "keySet". [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](http://home.kk-sc.de:9000/coding_rules#rule_key=squid%3AS2864)